src/gui/qgssnaptogridcanvasitem.cpp

```diff
@@ -0,0 +1,166 @@
+/***************************************************************************
+    qgssnaptogridcanvasitem.cpp
+    ----------------------
+    begin                : August 2018
+    copyright            : (C) Matthias Kuhn
+    email                : matthias@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgssnaptogridcanvasitem.h"
+#include "qgsmapcanvas.h"
+
+QgsSnapToGridCanvasItem::QgsSnapToGridCanvasItem( QgsMapCanvas *mapCanvas )
+  : QgsMapCanvasItem( mapCanvas )
+  , mGridPen( QPen( QColor( 127, 127, 127, 150 ), 1 ) )
+  , mCurrentPointPen( QPen( QColor( 200, 200, 200, 150 ), 3 ) )
+{
+  updateMapCanvasCrs();
+  connect( mMapCanvas, &QgsMapCanvas::extentsChanged, this, &QgsSnapToGridCanvasItem::updateZoomFactor );
+  connect( mMapCanvas, &QgsMapCanvas::destinationCrsChanged, this, &QgsSnapToGridCanvasItem::updateMapCanvasCrs );
+}
+
+void QgsSnapToGridCanvasItem::paint( QPainter *painter )
+{
+  painter->save();
+  QgsRectangle mapRect = mMapCanvas->extent();
+  if ( rect() != mapRect )
+    setRect( mapRect );
+
+  painter->setRenderHints( QPainter::Antialiasing );
+  painter->setCompositionMode( QPainter::CompositionMode_Difference );
+
+  if ( mEnabled && mAvailableByZoomFactor )
```

Comment:
I'd move this check up to the start of this function - otherwise you're needlessly saving/restoring the painter when you aren't drawing anything.